crypto/pkcs11key/pkcs11key.go

```diff
@@ -158,36 +131,76 @@ func (ps *PKCS11Key) Destroy() {
 	}
 }
 
-func (ps *PKCS11Key) openSession() (session pkcs11.SessionHandle, err error) {
+// Look up the token that contains the desired private key
+func (ps *PKCS11Key) openSession() (pkcs11.SessionHandle, pkcs11.ObjectHandle, error) {
+	var emptySession pkcs11.SessionHandle
+	var emptyHandle pkcs11.ObjectHandle
+
 	// Find slot by description
 	slots, err := ps.module.GetSlotList(true)
 	if err != nil {
-		return
+		return emptySession, emptyHandle, err
 	}
 	for _, slot := range slots {
-		slotInfo, err := ps.module.GetSlotInfo(slot)
-		if err != nil {
-			continue
+		// If ps.slotDescription is provided, only check matching slots
+		if ps.slotDescription != "" {
+			slotInfo, err := ps.module.GetSlotInfo(slot)
+			if err != nil {
+				continue
```

Comment:
There are a number of places in this loop where you continue if there's an error, but I think it would be more appropriate in each case to exit with that error.